src/SymfonyCache/TaggableStore.php

```diff
@@ -0,0 +1,387 @@
+                $request->headers
+            );
+
+            if ($varyKeyRequest === $varyKeyResponse) {
+                return $this->restoreResponse($responseData);
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Writes a cache entry to the store for the given Request and Response.
+     *
+     * Existing entries are read and any that match the response are removed. This
+     * method calls write with the new list of cache entries.
+     *
+     * @param Request  $request  A Request instance
+     * @param Response $response A Response instance
+     *
+     * @return string The key under which the response is stored
+     */
+    public function write(Request $request, Response $response)
+    {
+        if (!$response->headers->has('X-Content-Digest')) {
+            $contentDigest = $this->generateContentDigest($response);
+
+            if (false === $this->save($contentDigest, $response->getContent())) {
+                throw new \RuntimeException('Unable to store the entity.');
+            }
+
+            $response->headers->set('X-Content-Digest', $contentDigest);
+
+            if (!$response->headers->has('Transfer-Encoding')) {
+                $response->headers->set('Content-Length', strlen($response->getContent()));
+            }
+        }
+
+        $cacheKey = $this->getCacheKey($request);
+        $headers = $response->headers->all();
+        unset($headers['age']);
+
+        $item = $this->cache->getItem($cacheKey);
+
+        if (!$item->isHit()) {
+            $entries = [];
+        } else {
+            $entries = $item->get();
+        }
+
+        // Add or replace entry with current Vary header key
+        $entries[$this->getVaryKey($response->getVary(), $response->headers)] = [
+            'vary' => $response->getVary(),
+            'headers' => $headers,
+            'status' => $response->getStatusCode(),
+        ];
+
+        // If the response has a Vary header we remove the non-varying entry
+        if ($response->hasVary()) {
+            unset($entries[self::NON_VARYING_KEY]);
+        }
+
+        // Support tagging
+        $tags = [];
+        if ($response->headers->has($this->purgeTagsHeader)) {
+            $tags = explode(',', $response->headers->get($this->purgeTagsHeader));
+        }
+
+        $this->save($cacheKey, $entries, $tags);
+
+        return $cacheKey;
+    }
+
+    /**
+     * Invalidates all cache entries that match the request.
+     *
+     * @param Request $request A Request instance
+     */
+    public function invalidate(Request $request)
+    {
+        $cacheKey = $this->getCacheKey($request);
+
+        $this->cache->deleteItem($cacheKey);
+    }
+
+    /**
+     * Locks the cache for a given Request.
+     *
+     * @param Request $request A Request instance
+     *
+     * @return bool|string true if the lock is acquired, the path to the current lock otherwise
+     */
+    public function lock(Request $request)
```

Comment:
seems to  be not used

Reply:
What do you mean? It's part of the `StoreInterface`.

Reply:
Oops.. nevermind